### PR TITLE
Switch ON objfind QA

### PR DESCRIPTION
**Pull request by debora-pe**
As titled

This works only for multi-slit reduction.
Here is how the plot looks like
1Dextraction
.

### CODE REVIEW

File: pypeit/reduce.py
Comment (profxj) [inline pypeit/reduce.py]: maybe we can remove this TODO?

**Review by profxj — APPROVED** [review]
Looks great, thx!

File: pypeit/coadd2d.py
Comment (kbwestfall) [inline pypeit/coadd2d.py]: Does this mean it's no longer possible to toggled off the plots?

File: pypeit/core/extract.py
Comment (kbwestfall) [inline pypeit/core/extract.py]: I think the logic here is difficult.  I.e., someone could set `show_peaks=True`, but `debug_all=False` and not provide an `outfile_objprof` and then `pyplot` will have created something that never gets flushed, which will mess up the next plot being made.  It would be good to clean up how the QA is done for this function, but, barring that, I would put the `plt.close('all')` call outside the `if...elif` block.

**Review by kbwestfall — APPROVED** [review]
I would need to dig in more, but it feels like it's now very difficult to know how to toggle the QA.  I guess the point of the PR is to set it as always on, but it leaves hooks that may or may not work or even be relevant.  It may be more effort than it's worth to clean all this up in this PR (or it maybe easier than my brief impression), but we should raise an issue so that we're tracking it.

In any case, I'm approving now so as not to hold up moving forward.

File: pypeit/core/extract.py
Comment (jhennawi) [inline pypeit/core/extract.py:1478]: I'm also not following the logic here. Is the issue whether or not the plotting is interactive? Note that you can save the file to a figure, and also show it. In other words, you could simply have

plt.title(qa_title + ': Slit# {:d}'.format(specobj_dict['SLITID']))
if outfile_objprof is not None: 
    plt.savefig(outfile_objprof, dpi=400)
if debug_all: 
   viewer, ch = display.show_image(image*(thismask*inmask))
   plt.show()

plt.close('all)

File: pypeit/core/extract.py
Comment (jhennawi) [inline pypeit/core/extract.py]: So if outfile_objprof is not None, but show_peaks is False, then the plot does not get made. I think we need to think through the QA logic here. 

The point is we either want to do some full interactive plotting, or we want to make QA files and we should have the flexibility to do one, the other, or both.

File: pypeit/pypeit.py
Comment (jhennawi) [inline pypeit/pypeit.py]: Why are you disabling self.show? I think we still want that option, which as per my comment above can be achieved if we specify whether we want interactive QA, or the QA plot to disk.

File: pypeit/reduce.py
Comment (jhennawi) [inline pypeit/reduce.py]: Let's discuss these options. I think we want to have the option to have interactive plotting of both object finding steps since that is useful for debugging. Currently you are disabling that. This does not exclude the option to write out the result of the second object finding to a QA file, and so we should also have the option to do that.

File: pypeit/scripts/ql_keck_mosfire.py
Comment (jhennawi) [inline pypeit/scripts/ql_keck_mosfire.py]: Is there a reason why this has been disabled? Please do not remove the previous show_peaks functionality.

**Review by jhennawi — CHANGES_REQUESTED** [review]
Hi @deborape, 

Many thanks for getting this going. Please see my comments on the code below. 

Two major comments: 

1) You appear to be changing the previous interactive functionality, which is in my opinion not desirable. The previous interactive functionality to show the interactive object finding plot for the first and second object finding is not at odds with your goal of writing out the result of the second object finding plot to a QA file. I think we just need to think a little more about the parameters and the logic. 

2) I don't understand why this should only work for multislit. Have you tested it on echelle? Given that this is all at your fingertips now, whereas someone trying to implement echelle in the future would have to climb the learning curve you climbed again, I ask that you please try to implement echelle. If that for some unforeseen reason turns out to be complicated, then we can postpone. 

Thanks!
Joe

**Comment by debora-pe** [comment]
@jhennawi , @kbwestfall , @profxj  I have updated the work for this PR. Now it works for both multislit and echelle, so please take a look.

Tests pass.

```
Dev Suite
--- PYPEIT DEVELOPMENT SUITE PASSED 136/136 TESTS (Masters ignored) ---
Testing Started at 2021-12-10T16:58:30.344252
Testing Completed at 2021-12-11T03:36:42.841274
Total Time: 10:38:12.497022
```
Unit tests
```
pytest . -W ignore
=============================== test session starts ================================
platform darwin -- Python 3.8.5, pytest-6.1.2, py-1.9.0, pluggy-0.13.1
rootdir: /Users/dpelliccia/PypeIt/PypeIt, configfile: setup.cfg
collected 304 items                                                                

test_alignments.py .                                                         [  0%]
test_arc.py .                                                                [  0%]
test_archive.py ..                                                           [  1%]
test_arcimage.py ..                                                          [  1%]
test_biasframe.py ..                                                         [  2%]
test_bitmask.py ....                                                         [  3%]
test_bpmimage.py ..                                                          [  4%]
test_bspline.py ...............                                              [  9%]
test_buildimage.py ..                                                        [ 10%]
test_calibrations.py .......                                                 [ 12%]
test_chk_calibs.py ..                                                        [ 13%]
test_coadd.py .                                                              [ 13%]
test_collate_1d.py ............                                              [ 17%]
test_cooked.py .                                                             [ 17%]
test_coremeta.py .                                                           [ 18%]
test_datacontainer.py ........                                               [ 20%]
test_detector.py ...                                                         [ 21%]
test_display.py .                                                            [ 22%]
test_edgetrace.py .                                                          [ 22%]
test_fitting.py ....                                                         [ 23%]
test_flatfield.py .                                                          [ 24%]
test_flexure.py ..                                                           [ 24%]
test_flux.py ....                                                            [ 25%]
test_fluxspec.py .....                                                       [ 27%]
test_frametype.py ..                                                         [ 28%]
test_history.py ....                                                         [ 29%]
test_load.py .                                                               [ 29%]
test_load_images.py ................                                         [ 35%]
test_maskdesign_matching.py ...                                              [ 36%]
test_masterframe.py .                                                        [ 36%]
test_match.py ....                                                           [ 37%]
test_metadata.py ......                                                      [ 39%]
test_moment.py ......                                                        [ 41%]
test_msgs.py ..                                                              [ 42%]
test_opticalmodel.py ....                                                    [ 43%]
test_parse.py ...                                                            [ 44%]
test_pca.py ........                                                         [ 47%]
test_processimage.py ..                                                      [ 48%]
test_procimg.py .....                                                        [ 49%]
test_pydl.py .                                                               [ 50%]
test_pypeit.py ..                                                            [ 50%]
test_pypeitimage.py ..                                                       [ 51%]
test_pypeitpar.py ................................                           [ 61%]
test_pypeitsetup.py ........                                                 [ 64%]
test_qa.py .                                                                 [ 64%]
test_runpypeit.py ..                                                         [ 65%]
test_scienceimage.py ...                                                     [ 66%]
test_scripts.py .................                                            [ 72%]
test_setups.py ......................                                        [ 79%]
test_skysub.py ..                                                            [ 79%]
test_slitmask.py ...                                                         [ 80%]
test_slittrace.py ....                                                       [ 82%]
test_spec2dobj.py ......                                                     [ 84%]
test_specobj.py .......                                                      [ 86%]
test_specobjs.py .....                                                       [ 88%]
test_spectrographs.py ....................                                   [ 94%]
test_syncspec.py .                                                           [ 95%]
test_telluric.py ..                                                          [ 95%]
test_traceimage.py .                                                         [ 96%]
test_utils.py ...                                                            [ 97%]
test_vcorr.py ..                                                             [ 97%]
test_wavemodel.py .                                                          [ 98%]
test_wavetilts.py ....                                                       [ 99%]
test_wvcalib.py ..                                                           [100%]

========================= 304 passed in 1324.24s (0:22:04) =========================
```

**Comment by debora-pe** [comment]
Examples of the QA plots.

pos_DE 20170425 50487-dra11_DEIMOS_20170425T140121 014_D07_S0091_obj_prof
pos_cN20170331S0221-cN20170331S0217-pisco_D01_S0003_obj_prof

File: pypeit/core/extract.py
Comment (profxj) [inline pypeit/core/extract.py]: Can you briefly describe what these 3 keys are for?

File: pypeit/core/extract.py
Comment (profxj) [inline pypeit/core/extract.py]: Add `SLITORD_ID` too?

File: pypeit/reduce.py
Comment (profxj) [inline pypeit/reduce.py:1053]: Should it simply default to `False`?

File: pypeit/reduce.py
Comment (profxj) [inline pypeit/reduce.py]: as above, default `save_objfindQA` to `False`?

**Review by profxj — APPROVED** [review]
Just a few more comments from me.
Nothing all that substantial

File: pypeit/core/extract.py
Comment (jhennawi) [inline pypeit/core/extract.py]: My suggestion is to not pass in this dictionary and synthesise the filename here as you are doing. Instead run the qa.set_qa_filename code in the routine that calls this core objfind method to generate the filename.

File: pypeit/core/extract.py
Comment (jhennawi) [inline pypeit/core/extract.py]: Get rid of the dict.

File: pypeit/core/extract.py
Comment (jhennawi) [inline pypeit/core/extract.py]: See my comment below. You should only be passing in the QA outfile here, not this dictionary.

File: pypeit/core/extract.py
Comment (jhennawi) [inline pypeit/core/extract.py]: Here it may be necessary to pass in the dict. I'd prefer however that this code just take a string which acts as the prefix for the QA and then you append on the slitord_id to generate that filename in the loop over orders.

File: pypeit/reduce.py
Comment (jhennawi) [reply pypeit/reduce.py:1053]: I think it should default to True.

File: pypeit/reduce.py
Comment (jhennawi) [inline pypeit/reduce.py:1112]: Run the code to synthesise the filename here ,and then only pass the core method the filename. This has the benefit that you reduce dependencies and failure modes in the core objfind method. For example, a string cannot fail, but a call to qa.set_qa_filename can fail, and that dependency is completely unnecessary since you can generate the filename outside of that routine.

**Review by jhennawi — CHANGES_REQUESTED** [review]
Minor changes requested related to simplifying arguments to the core objfind and ech_objfind methods and also reducing dependencies in those methods. 

Can you also post an example of how the QA file hierarchy looks for echelle, and then test the 

run_pypeit -s by hand, since we don't have that in our current development tests.

File: pypeit/reduce.py
Comment (debora-pe) [reply pypeit/reduce.py:1053]: Now default to True.

File: pypeit/core/extract.py
Comment (debora-pe) [reply pypeit/core/extract.py]: Done

File: pypeit/core/extract.py
Comment (debora-pe) [reply pypeit/core/extract.py]: Done

File: pypeit/core/extract.py
Comment (debora-pe) [reply pypeit/core/extract.py]: Done

File: pypeit/core/extract.py
Comment (debora-pe) [reply pypeit/core/extract.py]: Done. It's a string and the `slitord_id` is appended in the loop over the orders.

File: pypeit/reduce.py
Comment (debora-pe) [reply pypeit/reduce.py:1112]: Done.

**Comment by debora-pe** [comment]
Hi @jhennawi! I addressed your comments and followed your suggestions. Please take a look when you can.
I'll run the tests today or tomorrow. 

I checked that `run_pypeit -s` still shows the objfind plots and it does. Also for `pypeit_coadd_2d`. 

Examples of the QA plots for an echelle and non-echelle spectrographs are above.
The filename of the QA plot includes the basename, the detector number and `slitord_id` (similarly to the spectral flexure correction plots).  E.g., "DE.20170425.50487-dra11_DEIMOS_20170425T140121.014_D07_S0034_obj_prof.png".
Also, if the a negative image exist, a prefix "neg_" or "pos_" is added to the file name. Eg.:
"neg_cN20170331S0216-pisco_GNIRS_20170331T085412.181_D01_S0003_obj_prof.png"
"pos_cN20170331S0216-pisco_GNIRS_20170331T085412.181_D01_S0003_obj_prof.png"

**Review by jhennawi — APPROVED** [review]
This looks great @deborape. Thanks for the very nice work.

**Comment by debora-pe** [comment]
Tests pass.
Dev Suite: 
```
--- PYPEIT DEVELOPMENT SUITE PASSED 136/136 TESTS (Masters ignored) ---
Testing Started at 2022-01-18T19:04:26.030363
Testing Completed at 2022-01-19T03:27:04.798528
Total Time: 8:22:38.768165
```

And unit tests
```
pytest . -W ignore
=============================== test session starts ===============================
platform darwin -- Python 3.8.5, pytest-6.1.2, py-1.9.0, pluggy-0.13.1
rootdir: /Users/dpelliccia/PypeIt/PypeIt, configfile: setup.cfg
collected 307 items                                                               

test_alignments.py .                                                        [  0%]
test_arc.py .                                                               [  0%]
test_archive.py ..                                                          [  1%]
test_arcimage.py ..                                                         [  1%]
test_biasframe.py ..                                                        [  2%]
test_bitmask.py ....                                                        [  3%]
test_bpmimage.py ..                                                         [  4%]
test_bspline.py ...............                                             [  9%]
test_buildimage.py ..                                                       [ 10%]
test_calibrations.py .......                                                [ 12%]
test_chk_calibs.py ..                                                       [ 13%]
test_coadd.py .                                                             [ 13%]
test_collate_1d.py ............                                             [ 17%]
test_cooked.py .                                                            [ 17%]
test_coremeta.py .                                                          [ 17%]
test_datacontainer.py ........                                              [ 20%]
test_detector.py ...                                                        [ 21%]
test_display.py .                                                           [ 21%]
test_edgetrace.py .                                                         [ 22%]
test_fitting.py ....                                                        [ 23%]
test_flatfield.py .                                                         [ 23%]
test_flexure.py ..                                                          [ 24%]
test_flux.py ....                                                           [ 25%]
test_fluxspec.py .....                                                      [ 27%]
test_frametype.py ..                                                        [ 28%]
test_history.py ....                                                        [ 29%]
test_load.py .                                                              [ 29%]
test_load_images.py ................                                        [ 34%]
test_manual.py ..                                                           [ 35%]
test_maskdesign_matching.py ...                                             [ 36%]
test_masterframe.py .                                                       [ 36%]
test_match.py ....                                                          [ 38%]
test_metadata.py ......                                                     [ 40%]
test_moment.py ......                                                       [ 42%]
test_msgs.py ..                                                             [ 42%]
test_opticalmodel.py ....                                                   [ 43%]
test_parse.py ...                                                           [ 44%]
test_pca.py ........                                                        [ 47%]
test_processimage.py ..                                                     [ 48%]
test_procimg.py .....                                                       [ 49%]
test_pydl.py .                                                              [ 50%]
test_pypeit.py ..                                                           [ 50%]
test_pypeitimage.py ..                                                      [ 51%]
test_pypeitpar.py ...............................                           [ 61%]
test_pypeitsetup.py ........                                                [ 64%]
test_qa.py .                                                                [ 64%]
test_runpypeit.py ..                                                        [ 65%]
test_scienceimage.py ...                                                    [ 66%]
test_scripts.py ..................                                          [ 71%]
test_setups.py ......................                                       [ 79%]
test_skysub.py ..                                                           [ 79%]
test_slitmask.py ...                                                        [ 80%]
test_slittrace.py ....                                                      [ 82%]
test_spec2dobj.py ......                                                    [ 84%]
test_specobj.py .......                                                     [ 86%]
test_specobjs.py .....                                                      [ 87%]
test_spectrographs.py ....................                                  [ 94%]
test_syncspec.py .                                                          [ 94%]
test_telluric.py ..                                                         [ 95%]
test_traceimage.py .                                                        [ 95%]
test_utils.py ....                                                          [ 97%]
test_vcorr.py ..                                                            [ 97%]
test_wavemodel.py .                                                         [ 98%]
test_wavetilts.py ....                                                      [ 99%]
test_wvcalib.py ..                                                          [100%]

======================== 307 passed in 1208.33s (0:20:08) =========================
```

**Comment by debora-pe** [comment]
merging

**Pull request merged**
